crates/libs/windows/src/core/event.rs

```diff
@@ -0,0 +1,254 @@
+            new_delegates.push(delegate);
+
+            let swap_lock = self.swap.lock().unwrap();
+            _lock_free_drop = self.delegates.swap(new_delegates);
+            token
+        })
+    }
+    /// Revokes a delegate's registration from the event object.
+    pub fn remove(&mut self, token: i64) -> Result<()> {
+        let mut _lock_free_drop = Array::new();
+        {
+            let change_lock = self.change.lock().unwrap();
+            if self.delegates.is_empty() {
+                return Ok(());
+            }
+            let mut capacity = self.delegates.len() - 1;
+            let mut new_delegates = Array::new();
+            let mut removed = false;
+            if capacity == 0 {
+                if self.delegates.as_slice()[0].to_token() == token {
+                    removed = true;
+                }
+            } else {
+                new_delegates = Array::with_capacity(capacity)?;
+                for delegate in self.delegates.as_slice() {
+                    if !removed && delegate.to_token() == token {
+                        removed = true;
+                        continue;
+                    }
+                    if capacity == 0 {
+                        debug_assert!(!removed);
+                        break;
+                    }
+                    new_delegates.push(delegate.clone());
+                    capacity -= 1;
+                }
+            }
+            if removed {
+                let swap_lock = self.swap.lock().unwrap();
+                _lock_free_drop = self.delegates.swap(new_delegates);
+            }
+        }
+        Ok(())
+    }
+    /// Clears the event, removing all delegates.
+    pub fn clear(&mut self) {
+        let mut _lock_free_drop = Array::new();
+        {
+            let change_lock = self.change.lock().unwrap();
+            if self.delegates.is_empty() {
+                return;
+            }
+            let swap_lock = self.swap.lock().unwrap();
+            _lock_free_drop = self.delegates.swap(Array::new());
+        }
+    }
+    /// Invokes all of the event object's registered delegates with the provided callback.
+    pub fn call<F: FnMut(&T) -> Result<()>>(&mut self, mut callback: F) -> Result<()> {
+        let lock_free_calls = {
+            let swap_lock = self.swap.lock().unwrap();
+            self.delegates.clone()
+        };
+        for delegate in lock_free_calls.as_slice() {
+            if let Err(error) = delegate.call(&mut callback) {
+                const RPC_E_SERVER_UNAVAILABLE: HRESULT = HRESULT(-2147023174); // HRESULT_FROM_WIN32(RPC_S_SERVER_UNAVAILABLE)
+                if matches!(error.code(), RPC_E_DISCONNECTED | JSCRIPT_E_CANTEXECUTE | RPC_E_SERVER_UNAVAILABLE) {
+                    self.remove(delegate.to_token())?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// A thread-safe reference-counted array of delegates.
+struct Array<T: Interface + Clone> {
```

Comment:
Nit: typically, the bounds on a generic param are only given where they're needed. So, here we would not constrain `T` but we would keep the constraint where it's really needed (on the inherit `impl` and the `Drop` impl).

Reply:
I tried that but Rust complains:

```
    Checking windows v0.37.0 (D:\git\windows-rs\crates\libs\windows)
error[E0367]: `Drop` impl requires `T: interface::Interface` but the struct it is implemented for does not
   --> crates\libs\windows\src\core\event.rs:190:9
    |
190 | impl<T: Interface + Clone> Drop for Array<T> {
    |         ^^^^^^^^^
    |
note: the implementor must specify the same requirement
   --> crates\libs\windows\src\core\event.rs:114:1
    |
114 | / struct Array<T> {
115 | |     buffer: *mut Buffer,
116 | |     len: usize,
117 | |     _phantom: std::marker::PhantomData<T>,
118 | | }
    | |_^

error[E0367]: `Drop` impl requires `T: Clone` but the struct it is implemented for does not
   --> crates\libs\windows\src\core\event.rs:190:21
    |
190 | impl<T: Interface + Clone> Drop for Array<T> {
    |                     ^^^^^
    |
note: the implementor must specify the same requirement
   --> crates\libs\windows\src\core\event.rs:114:1
    |
114 | / struct Array<T> {
115 | |     buffer: *mut Buffer,
116 | |     len: usize,
117 | |     _phantom: std::marker::PhantomData<T>,
118 | | }
    | |_^

For more information about this error, try `rustc --explain E0367`.
error: could not compile `windows` due to 2 previous errors
```